src/main/java/org/brapi/test/BrAPITestServer/service/germ/PedigreeService.java

```diff
@@ -176,22 +172,74 @@ public Optional<PedigreeNodeEntity> getPedigreeNode(String germplasmDbId) {
 		return node;
 	}
 
-	public PedigreeNodeEntity findOrCreatePedigreeNode(String germplasmDbId) throws BrAPIServerException {
-		Optional<PedigreeNodeEntity> nodeOpt = getPedigreeNode(germplasmDbId);
-		PedigreeNodeEntity node;
-		if (nodeOpt.isPresent()) {
-			node = nodeOpt.get();
-		} else {
-			GermplasmEntity germplasm = germplasmService.getGermplasmEntity(germplasmDbId);
+	public List<PedigreeNodeEntity> getPedigreeNodes(List<String> germplasmDbIds) {
+		List<PedigreeNodeEntity> nodes = new ArrayList<>();
+
+		// TODO: Might have to make a custom query for this that fetches the germ eagerly, bc need to compare the germIds.  Have to see if this is a significant performance hit.
+		List<PedigreeNodeEntity> dbNodeList = pedigreeRepository.findByGermplasm_IdIn(germplasmDbIds);
+
+		Map<String, List<PedigreeNodeEntity>> nodesGroupedByGerm = dbNodeList.stream().collect(Collectors.groupingBy(pn -> pn.getGermplasm().getId()));
+
+		nodesGroupedByGerm.forEach((germId, nodesByGerm) -> {
+			if (nodesByGerm.size() > 1) {
+				log.error("multiple pedigree nodes found for a single germplasm");
+			}
+
+			Optional<PedigreeNodeEntity> node = nodesByGerm.stream().findFirst();
+
+			node.ifPresent(nodes::add);
+		});
+
+		return nodes;
+	}
+
+	public List<PedigreeNodeEntity> findOrCreatePedigreeNodesFromGermplasmIds(List<String> germplasmDbIds) throws BrAPIServerException {
+
+        List<PedigreeNodeEntity> dbNodes = getPedigreeNodes(germplasmDbIds);
+
+        List<PedigreeNodeEntity> resultingNodes = new ArrayList<>(dbNodes);
+
+		// Find out which germIds were not found in the DB. Use a set for improved performance on the contains check.
+		// TODO: Check if the germEntity is already populated by getPedigreeNodes, and if this block results in more DB transactions.
+		Set<String> germIdsOfFoundNodes = dbNodes.stream()
+				.map(PedigreeNodeEntity::getGermplasm)
+				.map(BrAPIBaseEntity::getId)
+				.collect(Collectors.toSet());
+
+		List<String> germIdsWithNoPedigreeRecord = germplasmDbIds.stream()
+				.filter(gId -> !germIdsOfFoundNodes.contains(gId))
```

Comment:
GID has a particular meaning in DeltaBreed (and other systems), it's actually synonymous with the BrAPI `accessionNumber` field of germplasm.
```suggestion
				.filter(dbId -> !germIdsOfFoundNodes.contains(dbId))
```

Reply:
I think it's pretty obvious it refers to germplasm in a method dealing with looking up germplasm, not whatever you are referring to.  If we really want to be crazy specific we can say `germDbId`, but feels like overkill for a lambda filter obviously relating to germs.

Reply:
Yes, it's obvious that it is a germplasm field, but which field is not obvious, because GID is a synonym for the  germplasm's `accessionNumber`, while you're actually using the variable to refer to the germplasm's `dbId`.

Reply:
Changed.